File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        products = [1] * len(nums)
        right = 1
        left = 1

        for i, n in enumerate(nums):
            products[i] *= left
            left *= n

            products[-1-i] *= right
            right *= nums[-1-i]

        return products
        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] l = new int[n];
        int[] r = new int[n];
        int[] ans = new int[n];

        l[0] = 1;
        r[n - 1] = 1;

        for (int i = 1; i < n; i++) {
            l[i] = l[i - 1] * nums[i - 1];
            r[n - i - 1] = r[n - i] * nums[n - i];
        } // for

        for (int i = 0; i < n; i++) ans[i] = r[i] * l[i];

        return ans;
    } // productExceptSelf
} // Solution